Lib/subprocess.py

```diff
@@ -1245,11 +1245,39 @@ def _check_timeout(self, endtime, orig_timeout, stdout_seq, stderr_seq,
         """Convenience for checking if a timeout has expired."""
         if endtime is None:
             return
+
+        def translate_newlines_partial_output(data, encoding, errors):
+            # Handle decoding the data considering it may be truncated
+            # mid-codepoint (ignore the trailing partial codepoint).
+            # See https://github.com/python/cpython/issues/87597
+            try:
+                output = self._translate_newlines(data, encoding, errors)
+            except UnicodeDecodeError as exc:
+                if exc.end == len(data):
+                    output = self._translate_newlines(data[:exc.start],
+                                                      encoding,
+                                                      errors)
+                else:
+                    raise
+            return output
+
         if skip_check_and_raise or _time() > endtime:
+            if stdout_seq is not None:
+                stdout = b''.join(stdout_seq)
+                if self.text_mode:
+                    stdout = translate_newlines_partial_output(
+                            stdout, self.stdout.encoding, self.stdout.errors)
+            else:
+                stdout = None
```

Comment:
Maybe this whole pattern could go into the helper function? `join_and_translate_newlines`?

Reply:
I think the if/else around the None makes sense here as is, but the join and text_mode conditional could be done in-function.

basically

```python
def join_and_maybe_decode(output_seq, data, encoding, errors):
    output = b''.join(output_seq)
    if self.text_mode:
        try:
            ... # existing translate_newlines_partial_output code
```

and here you'd have things like

```python
if stdout_seq is not None:
    stdout = join_and_maybe_decode(stdout_seq, data, encoding, errors)
else:
    stdout = None

# and the same for stderr
```